### ISSUE TITLE
BufferedWaveProvider as a Stream Input?

### ISSUE THREAD

**Issue by Archetype90** [naudio/NAudio#415]
Is there a BufferedWaveProvider implementation that is a WaveStream or Stream? I realize that the entire purpose of IWaveProvider (which BWP inherits from) is to remove the concept of position from a Stream. I have attempted my own implementation of the BufferedWaveProvider that inherits from WaveStream, but I am simply too naive (thus far) to understand how to properly represent position of a CircularBuffer (which I think doesn't have a concept of position? At least like a MemoryStream does). 

To give more information on my use care - my goal is to use a BufferedWaveProvider (or something similar) as an input to [this Microsoft .NET speech recognition engine.](https://docs.microsoft.com/en-us/dotnet/api/system.speech.recognition.speechrecognitionengine.setinputtoaudiostream?view=netframework-4.7.2) The audio is coming in over the network. I have code now that will capture the audio and play it in real time (using a Blog post from @markheath) or code that will save it to a memory stream and do recognition on the memory stream after the processing is complete. But, I would like to be processing the BufferedWaveProvider as it is being read. 

I apologize that this is likely not the correct place for this question, but it is so specific to this library that I thought this would be a good place to get assistance and see if something like this exists.

**Comment by markheath** [comment]
I don't think you can actually stream audio into that recognition engine. It expects WAV input, and WAV files have a header that declares their length up front, making it unsuitable for live streaming

**Comment by rurounijones** [comment]
For reference for anyone coming across this issue (as I have) in the far future: 

While I still haven't actually figured out how to use it (New to audio stuff in general) the the Microsoft Cognitive Speech SDK now supports audio streams as input and output https://docs.microsoft.com/en-us/dotnet/api/microsoft.cognitiveservices.speech.audio?view=azure-dotnet

So you should be able to write a class that inherits from `PullAudioInputStreamCallback` that uses an NAudio `BufferedWaveProvider` in the background for example. Something along the lines of https://github.com/Azure-Samples/cognitive-services-speech-sdk/blob/bc67ba7ce88cffcf21dc96b04cb1b2835178c15c/samples/csharp/sharedcontent/console/helper.cs#L86
